diff --git a/usr/src/sys/arm/include/_types.h b/usr/src/sys/arm/include/_types.h
index 3012f97..6ba4bd3 100644
--- a/usr/src/sys/arm/include/_types.h
+++ b/usr/src/sys/arm/include/_types.h
@@ -88,7 +88,6 @@ typedef	__int64_t	__int_least64_t;
 typedef	__int32_t	__ptrdiff_t;		/* ptr1 - ptr2 */
 typedef	__int32_t	__register_t;
 typedef	__int32_t	__segsz_t;		/* segment size (in pages) */
-typedef	__uint32_t	__size_t;		/* sizeof() */
 typedef	__int32_t	__ssize_t;		/* byte count or error */
 typedef	__int64_t	__time_t;		/* time()... */
 typedef	__uint32_t	__uintfptr_t;
diff --git a/usr/src/sys/x86/include/_types.h b/usr/src/sys/x86/include/_types.h
index 07893c6..719e911 100644
--- a/usr/src/sys/x86/include/_types.h
+++ b/usr/src/sys/x86/include/_types.h
@@ -103,7 +103,6 @@ typedef	__int64_t	__int_least64_t;
 typedef	__int64_t	__ptrdiff_t;		/* ptr1 - ptr2 */
 typedef	__int64_t	__register_t;
 typedef	__int64_t	__segsz_t;		/* segment size (in pages) */
-typedef	__uint64_t	__size_t;		/* sizeof() */
 typedef	__int64_t	__ssize_t;		/* byte count or error */
 typedef	__int64_t	__time_t;		/* time()... */
 typedef	__uint64_t	__uintfptr_t;
@@ -112,7 +111,6 @@ typedef	__uint64_t	__uintptr_t;
 typedef	__int32_t	__ptrdiff_t;
 typedef	__int32_t	__register_t;
 typedef	__int32_t	__segsz_t;
-typedef	__uint32_t	__size_t;
 typedef	__int32_t	__ssize_t;
 typedef	__int32_t	__time_t;
 typedef	__uint32_t	__uintfptr_t;
